File: src/main/java/com/floorestimatepro/ui/MainFrame.java
package com.floorestimatepro.ui;

import java.awt.BorderLayout;
import java.awt.FlowLayout;
import java.io.File;
import javax.swing.JButton;
import javax.swing.JFileChooser;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTextField;
import javax.swing.filechooser.FileNameExtensionFilter;

import com.floorestimatepro.model.Calibration;
import com.floorestimatepro.model.EstimateCalculator;
import com.floorestimatepro.model.EstimateResult;
import com.floorestimatepro.model.PlanPoint;
import com.floorestimatepro.model.Project;
import com.floorestimatepro.persistence.ProjectStore;
import com.floorestimatepro.export.EstimateExporter;

public class MainFrame extends JFrame{
    Project project = new Project();
    PlanCanvas canvas = new PlanCanvas(project);
    JTextField knownFeetField = new JTextField("10", 5);
    JTextField wasteField = new JTextField("1.10", 5);
    JTextField priceField = new JTextField("5.0", 5);
    JLabel statusLabel = new JLabel("Open a plan image to start");
    JLabel resultLabel = new JLabel("Estimate: —");

    public MainFrame(){
        setTitle("Floor Estimate Pro");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLayout(new BorderLayout());

        JPanel toolbar = new JPanel(new FlowLayout(FlowLayout.LEFT));
        JPanel toolbar2 = new JPanel(new FlowLayout(FlowLayout.LEFT));
        JButton openBtn = new JButton("Open Image");
        JButton calibrateBtn = new JButton("Calibrate");
        JButton applyCalBtn = new JButton("Apply Calibration");
        JButton roomBtn = new JButton("Draw Room");
        JButton finishRoomBtn = new JButton("Finish Room");
        JButton obstacleBtn = new JButton("Draw Obstacle");
        JButton finishObstacleBtn = new JButton("Finish Obstacle");
        JButton estimateBtn = new JButton("Estimate");
        JButton finishEstimateBtn = new JButton("Finish Estimate");
        JButton saveBtn = new JButton("Save");
        JButton loadBtn = new JButton("Load");
        JButton exportBtn = new JButton("Export");
        JButton zoomInBtn = new JButton("Zoom In");
        JButton zoomOutBtn = new JButton("Zoom Out");
        JButton zoomResetBtn = new JButton("Zoom 100%");

        toolbar.add(openBtn);
        toolbar.add(zoomInBtn);
        toolbar.add(zoomOutBtn);
        toolbar.add(zoomResetBtn);
        toolbar.add(calibrateBtn);
        toolbar.add(new JLabel("Feet:"));
        toolbar.add(knownFeetField);
        toolbar.add(applyCalBtn);
        toolbar.add(roomBtn);
        toolbar.add(finishRoomBtn);
        toolbar.add(obstacleBtn);
        toolbar.add(finishObstacleBtn);

        toolbar2.add(new JLabel("Waste:"));
        toolbar2.add(wasteField);
        toolbar2.add(new JLabel("Price:"));
        toolbar2.add(priceField);
        toolbar2.add(estimateBtn);
        toolbar2.add(finishEstimateBtn);
        toolbar2.add(saveBtn);
        toolbar2.add(loadBtn);
        toolbar2.add(exportBtn);

        JPanel north = new JPanel(new BorderLayout());
        north.add(toolbar, BorderLayout.NORTH);
        north.add(toolbar2, BorderLayout.SOUTH);

        JPanel south = new JPanel(new BorderLayout());
        south.add(statusLabel, BorderLayout.WEST);
        south.add(resultLabel, BorderLayout.EAST);

        add(north, BorderLayout.NORTH);
        add(new JScrollPane(canvas), BorderLayout.CENTER);
        add(south, BorderLayout.SOUTH);

        canvas.setOnChanged(this::refreshStatus);

        openBtn.addActionListener(e -> onOpenImage());
        zoomInBtn.addActionListener(e -> canvas.zoomIn());
        zoomOutBtn.addActionListener(e -> canvas.zoomOut());
        zoomResetBtn.addActionListener(e -> canvas.resetZoom());
        calibrateBtn.addActionListener(e -> onCalibrateTool());
        applyCalBtn.addActionListener(e -> onApplyCalibration());
        roomBtn.addActionListener(e -> onDrawRoomTool());
        finishRoomBtn.addActionListener(e -> onFinishRoom());
        obstacleBtn.addActionListener(e -> onDrawObstacleTool());
        finishObstacleBtn.addActionListener(e -> onFinishObstacle());
        estimateBtn.addActionListener(e -> refreshEstimate());
        finishEstimateBtn.addActionListener(e -> onFinishEstimate());
        saveBtn.addActionListener(e -> onSave());
        loadBtn.addActionListener(e -> onLoad());
        exportBtn.addActionListener(e -> onExport());

        pack();
        setLocationRelativeTo(null);
    }

    void onOpenImage(){
        JFileChooser chooser = new JFileChooser();
        chooser.setFileFilter(new FileNameExtensionFilter("Images", "png", "jpg", "jpeg", "gif"));
        if(chooser.showOpenDialog(this) != JFileChooser.APPROVE_OPTION){
            return;
        }
        File file = chooser.getSelectedFile();
        project.setImagePath(file.getAbsolutePath());
        canvas.loadImageFromProject();
        statusLabel.setText("Loaded " + file.getName());
    }

    void onCalibrateTool(){
        canvas.setToolMode(ToolMode.CALIBRATE);
        statusLabel.setText("Click two ends of a known length, enter feet, then Apply Calibration");
    }

    void onApplyCalibration(){
        if(canvas.pendingPts().size() < 2){
            JOptionPane.showMessageDialog(this, "Click two calibration points first");
            return;
        }
        try{
            double knownFeet = Double.parseDouble(knownFeetField.getText().trim());
            PlanPoint a = canvas.pendingPts().get(0);
            PlanPoint b = canvas.pendingPts().get(1);
            Calibration cal = Calibration.fromKnownLength(a, b, knownFeet);
            project.setCalibration(cal);
            canvas.pendingPts().clear();
            canvas.repaint();
            statusLabel.setText("Calibrated: " + cal.feetPerPixel() + " ft/px");
            refreshEstimate();
        }catch(Exception ex){
            JOptionPane.showMessageDialog(this, ex.getMessage());
        }
    }

    void onDrawRoomTool(){
        canvas.setToolMode(ToolMode.DRAW_ROOM);
        statusLabel.setText("Click room corners, then Finish Room");
    }

    void onFinishRoom(){
        try{
            canvas.finishRoom();
            statusLabel.setText("Room added (" + project.roomCount() + ")");
            refreshEstimate();
        }catch(Exception ex){
            JOptionPane.showMessageDialog(this, ex.getMessage());
        }
    }

    void onDrawObstacleTool(){
        canvas.setToolMode(ToolMode.DRAW_OBSTACLE);
        statusLabel.setText("Click obstacle corners, then Finish Obstacle");
    }

    void onFinishObstacle(){
        try{
            canvas.finishObstacle();
            statusLabel.setText("Obstacle added (" + project.obstacleCount() + ")");
            refreshEstimate();
        }catch(Exception ex){
            JOptionPane.showMessageDialog(this, ex.getMessage());
        }
    }

    void refreshStatus(){
        String zoomText = " | Zoom: " + Math.round(canvas.getZoom() * 100) + "%";
        if(canvas.getToolMode() == ToolMode.CALIBRATE){
            statusLabel.setText("Calibration clicks: " + canvas.pendingPts().size() + "/2" + zoomText);
        }else if(canvas.getToolMode() == ToolMode.DRAW_ROOM){
            statusLabel.setText("Room corners: " + canvas.pendingPts().size() + zoomText);
        }else if(canvas.getToolMode() == ToolMode.DRAW_OBSTACLE){
            statusLabel.setText("Obstacle corners: " + canvas.pendingPts().size() + zoomText);
        }else{
            statusLabel.setText("Zoom: " + Math.round(canvas.getZoom() * 100) + "%");
        }
    }

    void refreshEstimate(){
        if(!project.isCalibrated()){
            resultLabel.setText("Estimate: calibrate first");
            return;
        }
        if(project.rooms().isEmpty()){
            resultLabel.setText("Estimate: add a room first");
            return;
        }
        try{
            double waste = Double.parseDouble(wasteField.getText().trim());
            double price = Double.parseDouble(priceField.getText().trim());
            EstimateResult result = EstimateCalculator.calculate(
                project.rooms(),
                project.obstacles(),
                project.calibration(),
                waste,
                price
            );
            resultLabel.setText(
                "Net px2: " + round(result.netPixelArea())
                + " | Sq ft: " + round(result.netRealArea())
                + " | Material: " + round(result.materialSqFt())
                + " | Cost: $" + round(result.estimatedCost())
            );
        }catch(Exception ex){
            resultLabel.setText("Estimate error: " + ex.getMessage());
        }
    }

    void onSave(){
        JFileChooser chooser = new JFileChooser();
        chooser.setFileFilter(new FileNameExtensionFilter("JSON", "json"));
        if(chooser.showSaveDialog(this) != JFileChooser.APPROVE_OPTION){
            return;
        }
        try{
            String path = chooser.getSelectedFile().getAbsolutePath();
            if(!path.endsWith(".json")){
                path = path + ".json";
            }
            ProjectStore.save(project, path);
            statusLabel.setText("Saved " + path);
        }catch(Exception ex){
            JOptionPane.showMessageDialog(this, ex.getMessage());
        }
    }

    void onLoad(){
        JFileChooser chooser = new JFileChooser();
        chooser.setFileFilter(new FileNameExtensionFilter("JSON", "json"));
        if(chooser.showOpenDialog(this) != JFileChooser.APPROVE_OPTION){
            return;
        }
        try{
            project = ProjectStore.load(chooser.getSelectedFile().getAbsolutePath());
            canvas.setProject(project);
            statusLabel.setText("Loaded project");
            refreshEstimate();
        }catch(Exception ex){
            JOptionPane.showMessageDialog(this, ex.getMessage());
        }
    }

    void onExport(){
        if(!project.isCalibrated() || project.rooms().isEmpty()){
            JOptionPane.showMessageDialog(this, "Calibrate and add a room before exporting");
            return;
        }
        JFileChooser chooser = new JFileChooser();
        chooser.setFileFilter(new FileNameExtensionFilter("CSV", "csv"));
        if(chooser.showSaveDialog(this) != JFileChooser.APPROVE_OPTION){
            return;
        }
        try{
            double waste = Double.parseDouble(wasteField.getText().trim());
            double price = Double.parseDouble(priceField.getText().trim());
            EstimateResult result = EstimateCalculator.calculate(
                project.rooms(),
                project.obstacles(),
                project.calibration(),
                waste,
                price
            );
            String path = chooser.getSelectedFile().getAbsolutePath();
            if(!path.endsWith(".csv")){
                path = path + ".csv";
            }
            EstimateExporter.exportCsv(result, path);
            statusLabel.setText("Exported " + path);
        }catch(Exception ex){
            JOptionPane.showMessageDialog(this, ex.getMessage());
        }
    }

    void onFinishEstimate(){
        if(!project.isCalibrated()){
            JOptionPane.showMessageDialog(this, "Calibrate first");
            return;
        }
        if(project.rooms().isEmpty()){
            JOptionPane.showMessageDialog(this, "Add a room first");
            return;
        }
        try{
            double waste = Double.parseDouble(wasteField.getText().trim());
            double price = Double.parseDouble(priceField.getText().trim());
            EstimateResult result = EstimateCalculator.calculate(
                project.rooms(),
                project.obstacles(),
                project.calibration(),
                waste,
                price
            );
            EstimateResultsFrame frame = new EstimateResultsFrame(
                result,
                project.roomCount(),
                project.obstacleCount()
            );
            frame.setVisible(true);
        }catch(Exception ex){
            JOptionPane.showMessageDialog(this, ex.getMessage());
        }
    }

    double round(double value){
        return Math.round(value * 1000.0) / 1000.0;
    }
}

File: python/src/floor_estimate_pro/ui/main_frame.py
from __future__ import annotations

from PySide6.QtWidgets import (
    QFileDialog,
    QHBoxLayout,
    QLabel,
    QLineEdit,
    QMainWindow,
    QMessageBox,
    QScrollArea,
    QVBoxLayout,
    QWidget,
    QPushButton,
)

from floor_estimate_pro.export.estimate_exporter import EstimateExporter
from floor_estimate_pro.model.calibration import Calibration
from floor_estimate_pro.model.estimate_calculator import EstimateCalculator
from floor_estimate_pro.model.project import Project
from floor_estimate_pro.persistence.project_store import ProjectStore
from floor_estimate_pro.ui.estimate_results_frame import EstimateResultsFrame
from floor_estimate_pro.ui.plan_canvas import PlanCanvas
from floor_estimate_pro.ui.tool_mode import ToolMode


class MainFrame(QMainWindow):
    def __init__(self) -> None:
        super().__init__()
        self.project = Project()
        self.canvas = PlanCanvas(self.project)
        self.known_feet_field = QLineEdit("10")
        self.known_feet_field.setFixedWidth(50)
        self.waste_field = QLineEdit("1.10")
        self.waste_field.setFixedWidth(50)
        self.price_field = QLineEdit("5.0")
        self.price_field.setFixedWidth(50)
        self.status_label = QLabel("Open a plan image to start")
        self.result_label = QLabel("Estimate: —")

        self.setWindowTitle("Floor Estimate Pro")

        open_btn = QPushButton("Open Image")
        calibrate_btn = QPushButton("Calibrate")
        apply_cal_btn = QPushButton("Apply Calibration")
        room_btn = QPushButton("Draw Room")
        finish_room_btn = QPushButton("Finish Room")
        obstacle_btn = QPushButton("Draw Obstacle")
        finish_obstacle_btn = QPushButton("Finish Obstacle")
        estimate_btn = QPushButton("Estimate")
        finish_estimate_btn = QPushButton("Finish Estimate")
        save_btn = QPushButton("Save")
        load_btn = QPushButton("Load")
        export_btn = QPushButton("Export")
        zoom_in_btn = QPushButton("Zoom In")
        zoom_out_btn = QPushButton("Zoom Out")
        zoom_reset_btn = QPushButton("Zoom 100%")

        toolbar = QHBoxLayout()
        toolbar.addWidget(open_btn)
        toolbar.addWidget(zoom_in_btn)
        toolbar.addWidget(zoom_out_btn)
        toolbar.addWidget(zoom_reset_btn)
        toolbar.addWidget(calibrate_btn)
        toolbar.addWidget(QLabel("Feet:"))
        toolbar.addWidget(self.known_feet_field)
        toolbar.addWidget(apply_cal_btn)
        toolbar.addWidget(room_btn)
        toolbar.addWidget(finish_room_btn)
        toolbar.addWidget(obstacle_btn)
        toolbar.addWidget(finish_obstacle_btn)
        toolbar.addStretch()

        toolbar2 = QHBoxLayout()
        toolbar2.addWidget(QLabel("Waste:"))
        toolbar2.addWidget(self.waste_field)
        toolbar2.addWidget(QLabel("Price:"))
        toolbar2.addWidget(self.price_field)
        toolbar2.addWidget(estimate_btn)
        toolbar2.addWidget(finish_estimate_btn)
        toolbar2.addWidget(save_btn)
        toolbar2.addWidget(load_btn)
        toolbar2.addWidget(export_btn)
        toolbar2.addStretch()

        south = QHBoxLayout()
        south.addWidget(self.status_label)
        south.addStretch()
        south.addWidget(self.result_label)

        scroll = QScrollArea()
        scroll.setWidget(self.canvas)
        scroll.setWidgetResizable(False)

        central = QWidget()
        root = QVBoxLayout(central)
        root.addLayout(toolbar)
        root.addLayout(toolbar2)
        root.addWidget(scroll, 1)
        root.addLayout(south)
        self.setCentralWidget(central)

        self.canvas.changed.connect(self.refresh_status)

        open_btn.clicked.connect(self.on_open_image)
        zoom_in_btn.clicked.connect(self.canvas.zoom_in)
        zoom_out_btn.clicked.connect(self.canvas.zoom_out)
        zoom_reset_btn.clicked.connect(self.canvas.reset_zoom)
        calibrate_btn.clicked.connect(self.on_calibrate_tool)
        apply_cal_btn.clicked.connect(self.on_apply_calibration)
        room_btn.clicked.connect(self.on_draw_room_tool)
        finish_room_btn.clicked.connect(self.on_finish_room)
        obstacle_btn.clicked.connect(self.on_draw_obstacle_tool)
        finish_obstacle_btn.clicked.connect(self.on_finish_obstacle)
        estimate_btn.clicked.connect(self.refresh_estimate)
        finish_estimate_btn.clicked.connect(self.on_finish_estimate)
        save_btn.clicked.connect(self.on_save)
        load_btn.clicked.connect(self.on_load)
        export_btn.clicked.connect(self.on_export)

        self.resize(1100, 750)

    def on_open_image(self) -> None:
        path, _ = QFileDialog.getOpenFileName(
            self, "Open Image", "", "Images (*.png *.jpg *.jpeg *.gif)"
        )
        if not path:
            return
        self.project.set_image_path(path)
        self.canvas.load_image_from_project()
        self.status_label.setText("Loaded " + path.split("/")[-1].split("\\")[-1])

    def on_calibrate_tool(self) -> None:
        self.canvas.set_tool_mode(ToolMode.CALIBRATE)
        self.status_label.setText(
            "Click two ends of a known length, enter feet, then Apply Calibration"
        )

    def on_apply_calibration(self) -> None:
        if len(self.canvas.pending_pts) < 2:
            QMessageBox.warning(self, "Calibration", "Click two calibration points first")
            return
        try:
            known_feet = float(self.known_feet_field.text().strip())
            a = self.canvas.pending_pts[0]
            b = self.canvas.pending_pts[1]
            cal = Calibration.from_known_length(a, b, known_feet)
            self.project.set_calibration(cal)
            self.canvas.pending_pts.clear()
            self.canvas.update()
            self.status_label.setText("Calibrated: " + str(cal.feet_per_pixel) + " ft/px")
            self.refresh_estimate()
        except Exception as ex:
            QMessageBox.warning(self, "Error", str(ex))

    def on_draw_room_tool(self) -> None:
        self.canvas.set_tool_mode(ToolMode.DRAW_ROOM)
        self.status_label.setText("Click room corners, then Finish Room")

    def on_finish_room(self) -> None:
        try:
            self.canvas.finish_room()
            self.status_label.setText("Room added (" + str(self.project.room_count()) + ")")
            self.refresh_estimate()
        except Exception as ex:
            QMessageBox.warning(self, "Error", str(ex))

    def on_draw_obstacle_tool(self) -> None:
        self.canvas.set_tool_mode(ToolMode.DRAW_OBSTACLE)
        self.status_label.setText("Click obstacle corners, then Finish Obstacle")

    def on_finish_obstacle(self) -> None:
        try:
            self.canvas.finish_obstacle()
            self.status_label.setText(
                "Obstacle added (" + str(self.project.obstacle_count()) + ")"
            )
            self.refresh_estimate()
        except Exception as ex:
            QMessageBox.warning(self, "Error", str(ex))

    def refresh_status(self) -> None:
        zoom_text = " | Zoom: " + str(round(self.canvas.get_zoom() * 100)) + "%"
        if self.canvas.get_tool_mode() == ToolMode.CALIBRATE:
            self.status_label.setText(
                "Calibration clicks: " + str(len(self.canvas.pending_pts)) + "/2" + zoom_text
            )
        elif self.canvas.get_tool_mode() == ToolMode.DRAW_ROOM:
            self.status_label.setText(
                "Room corners: " + str(len(self.canvas.pending_pts)) + zoom_text
            )
        elif self.canvas.get_tool_mode() == ToolMode.DRAW_OBSTACLE:
            self.status_label.setText(
                "Obstacle corners: " + str(len(self.canvas.pending_pts)) + zoom_text
            )
        else:
            self.status_label.setText("Zoom: " + str(round(self.canvas.get_zoom() * 100)) + "%")

    def refresh_estimate(self) -> None:
        if not self.project.is_calibrated():
            self.result_label.setText("Estimate: calibrate first")
            return
        if len(self.project.rooms()) == 0:
            self.result_label.setText("Estimate: add a room first")
            return
        try:
            waste = float(self.waste_field.text().strip())
            price = float(self.price_field.text().strip())
            result = EstimateCalculator.calculate(
                self.project.rooms(),
                self.project.obstacles(),
                self.project.calibration(),
                waste,
                price,
            )
            self.result_label.setText(
                "Net px2: " + str(self.round(result.net_pixel_area))
                + " | Sq ft: " + str(self.round(result.net_real_area))
                + " | Material: " + str(self.round(result.material_sq_ft))
                + " | Cost: $" + str(self.round(result.estimated_cost))
            )
        except Exception as ex:
            self.result_label.setText("Estimate error: " + str(ex))

    def on_save(self) -> None:
        path, _ = QFileDialog.getSaveFileName(self, "Save", "", "JSON (*.json)")
        if not path:
            return
        try:
            if not path.endswith(".json"):
                path = path + ".json"
            ProjectStore.save(self.project, path)
            self.status_label.setText("Saved " + path)
        except Exception as ex:
            QMessageBox.warning(self, "Error", str(ex))

    def on_load(self) -> None:
        path, _ = QFileDialog.getOpenFileName(self, "Load", "", "JSON (*.json)")
        if not path:
            return
        try:
            self.project = ProjectStore.load(path)
            self.canvas.set_project(self.project)
            self.status_label.setText("Loaded project")
            self.refresh_estimate()
        except Exception as ex:
            QMessageBox.warning(self, "Error", str(ex))

    def on_export(self) -> None:
        if not self.project.is_calibrated() or len(self.project.rooms()) == 0:
            QMessageBox.warning(self, "Export", "Calibrate and add a room before exporting")
            return
        path, _ = QFileDialog.getSaveFileName(self, "Export CSV", "", "CSV (*.csv)")
        if not path:
            return
        try:
            waste = float(self.waste_field.text().strip())
            price = float(self.price_field.text().strip())
            result = EstimateCalculator.calculate(
                self.project.rooms(),
                self.project.obstacles(),
                self.project.calibration(),
                waste,
                price,
            )
            if not path.endswith(".csv"):
                path = path + ".csv"
            EstimateExporter.export_csv(result, path)
            self.status_label.setText("Exported " + path)
        except Exception as ex:
            QMessageBox.warning(self, "Error", str(ex))

    def on_finish_estimate(self) -> None:
        if not self.project.is_calibrated():
            QMessageBox.warning(self, "Estimate", "Calibrate first")
            return
        if len(self.project.rooms()) == 0:
            QMessageBox.warning(self, "Estimate", "Add a room first")
            return
        try:
            waste = float(self.waste_field.text().strip())
            price = float(self.price_field.text().strip())
            result = EstimateCalculator.calculate(
                self.project.rooms(),
                self.project.obstacles(),
                self.project.calibration(),
                waste,
                price,
            )
            frame = EstimateResultsFrame(
                result,
                self.project.room_count(),
                self.project.obstacle_count(),
            )
            frame.exec()
        except Exception as ex:
            QMessageBox.warning(self, "Error", str(ex))

    def round(self, value: float) -> float:
        return round(value * 1000.0) / 1000.0
